src/utils/decorators/input-validation/input-validation.js

```diff
@@ -263,8 +264,19 @@ const InputValidation = ComposedComponent => class Component extends ComposedCom
         // figure out if the message is positioned offscreen
         const messageScreenPosition = message.getBoundingClientRect().left + message.offsetWidth;
 
+        let shouldFlip = false;
+
         // change the position if it is offscreen
-        if (messageScreenPosition > this._window.innerWidth) {
+        if (this.context.modal && this.context.modal.getDialog()) {
+          // if in a modal check its position relative to that
+          const dialog = this.context.modal.getDialog();
+          shouldFlip = (message.offsetLeft + this._target.offsetLeft + message.offsetWidth) > dialog.offsetWidth;
+        } else {
+          // otherwise check relative to the window
+          shouldFlip = messageScreenPosition > this._window.innerWidth;
+        }
+
+        if (shouldFlip) {
           messagePositionLeft -= message.offsetWidth;
           message.style.left = `${messagePositionLeft}px`;
```

Comment:
Might get better performance if we cache things like `message.offsetWidth` and `this._window.innerWidth` in local variables, rather than repeatedly reading them.